crates/curp/src/server/cmd_worker/mod.rs

```diff
@@ -36,29 +36,37 @@ where
     }
 }
 
+/// ER and ASR
+type ErAsr<C> = (<C as Command>::ER, Option<<C as Command>::ASR>);
+
 /// Cmd worker execute handler
 pub(super) fn execute<C: Command, CE: CommandExecutor<C>, RC: RoleChange>(
     entry: &LogEntry<C>,
     ce: &CE,
     curp: &RawCurp<C, RC>,
-) -> Result<<C as Command>::ER, <C as Command>::Error> {
+) -> Result<ErAsr<C>, <C as Command>::Error> {
+    let cb = curp.cmd_board();
     let id = curp.id();
-    match entry.entry_data {
-        EntryData::Command(ref cmd) => {
-            let er = ce.execute(cmd);
-            debug!(
-                "{id} cmd({}) is speculatively executed, exe status: {}",
-                entry.propose_id,
-                er.is_ok(),
-            );
-            er
-        }
-        EntryData::ConfChange(_)
-        | EntryData::Shutdown
-        | EntryData::Empty
-        | EntryData::SetNodeState(_, _, _) => {
-            unreachable!("should not speculative execute {:?}", entry.entry_data)
-        }
+    let EntryData::Command(ref cmd) = entry.entry_data else {
+        unreachable!("should not speculative execute {:?}", entry.entry_data);
+    };
+    if cmd.is_read_only() {
+        let result = ce
+            .after_sync(vec![AfterSyncCmd::new(cmd, true)], None)
+            .remove(0)?;
+        let (asr, er_opt) = result.into_parts();
+        let er = er_opt.unwrap_or_else(|| unreachable!("er should exist"));
+        Ok((er, Some(asr)))
```

Comment:
little suggestion: `AfterSyncOk::into_parts` returns the tuple in form of `(asr, er)`, but `ErAsr` is a `(er, asr)`.
I think it maybe better to unify them into one form.

Reply:
They are different types. The first is `(Option<C::ASR>, C::ER)` and the second is `(C::ER, Option<C::ASR>)`